### PR TITLE
Raise specific error at typing time for iteration on >1D array.

### LINKED ISSUE #4508: LoweringError iterating over 2D array @numba.jit(nopython=True)
#### # FAILS: **This should not have happened**, a problem has occurred in Numba's internals.  You are currently using Numba version 0.45.1...LoweringError: Failed in nopython mode pipeline (step: nopython mode backend) iterating over 2D array...**Please report**...
a=numpy.array([[0, 1],[1, 3],[2, 2]])
@numba.jit(nopython=True)
def f(a):
..for i in a:
....print(i)
f(a)

#### # RUNS WITH A WARNING: Compilation is falling back to object mode WITH looplifting enabled because Function f failed at nopython mode lowering due to: iterating over 2D array
a=numpy.array([[0, 1],[1, 3],[2, 2]])
@numba.jit()
def f(a):
.. for i in a:
....print(i)
f(a)

### I figured out that the way to do what I need:
a=numpy.array([[0, 1],[1, 3],[2, 2]])
@numba.jit(nopython=True, nogil=True, cache=True)
def f(a):
..lst=[]
..for i in range(0,len(a)):
....lst.append(a[i,1])
..return lst(a)
f(a)
### filing the bug because is was requested by @numba.jit error message

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #4508

### CODE REVIEW

File: numba/typing/builtins.py [hunk 1]
```diff
@@ -88,7 +88,16 @@ def generic(self, args, kws):
         assert not kws
         [obj] = args
         if isinstance(obj, types.IterableType):
-            return signature(obj.iterator_type, obj)
+            # raise this here to provide a very specific message about this
```
Comment (esc) [inline numba/typing/builtins.py]:
```suggestion
            # Raise this here to provide a very specific message about this
```

File: numba/typing/builtins.py [hunk 2]
```diff
@@ -88,7 +88,16 @@ def generic(self, args, kws):
...
+            # common issue, delaying the error until later leads to something
+            # less specific being noted as the problem (e.g. no support for
+            # getiter on array(<>, 2, 'C')).
+            if isinstance(obj, types.Array) and obj.ndim > 1:
+                msg = ("Direct iteration is not supported for arrays with "
+                       "dimension > 1. Try using indexing instead?")
```
Comment (esc) [inline numba/typing/builtins.py]:
```suggestion
                       "dimension > 1. Try using indexing instead.")
```

**Comment by esc** [comment]
@stuartarchibald thanks for the fixes, can be marked for merge pending CI being green.

**Comment by stuartarchibald** [comment]
CI Passed, thanks for the review.

**Pull request merged**